### PR TITLE
extend monitoring by adding metrics for adapters

### PR DESCRIPTION
**Pull request by bossenti**
### Purpose

While we already expose monitoring metrics for pipelines during the pipeline health check, adapter metrics are missing out so far.
This PR introduces the capability to define metrics for adapters (see `AdapterMetrics`) and expose them via Prometheus during the adapter health check.
For now, this supports just one metric (`adapter_events_published_total`) which tracks the total amount of events published for every adapter.
This can be used, for example, to monitor if all adapters are running.
Here is a screenshot of the metric for two adapters in Prometheus:

image

### Remarks

PR introduces (a) breaking change(s): no

PR introduces (a) deprecation(s): no

### CODE REVIEW

File: streampipes-commons/src/test/java/org/apache/streampipes/commons/prometheus/adapter/AdapterMetricsTest.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+import static junit.framework.TestCase.assertTrue;
+import static org.junit.Assert.assertThrows;
+
+public class AdapterMetricsTest {
+
+  @Test
+  public void register() {
+
+    var metrics = new AdapterMetrics();
+    var adapterId = "adapterId";
```
Comment (tenthe) [inline streampipes-commons/src/test/java/org/apache/streampipes/commons/prometheus/adapter/AdapterMetricsTest.java:35]: Can we use `adapterId` & `adapterNam` as class variables and use it for all the three tests?

**Review by tenthe — CHANGES_REQUESTED** [review]

File: streampipes-connect-management/src/main/java/org/apache/streampipes/connect/management/management/AdapterMasterManagement.java [hunk 2]
```diff
@@ -166,6 +171,9 @@ public void startStreamAdapter(String elementId) throws AdapterException {
       // Invoke adapter instance
       WorkerRestClient.invokeStreamAdapter(baseUrl, elementId);
 
+      // register the adapter at the metrics manager so that the AdapterHealthCheck can send metrics
+      AdapterMetricsManager.INSTANCE.getAdapterMetrics().register(ad.getElementId(), ad.getName());
```
[the remark this answers is not in the document]
Comment (bossenti) [reply streampipes-connect-management/src/main/java/org/apache/streampipes/connect/management/management/AdapterMasterManagement.java:173]: Done

File: streampipes-connect-management/src/main/java/org/apache/streampipes/connect/management/management/AdapterMasterManagement.java [hunk 3]
```diff
@@ -149,6 +150,10 @@ public void stopStreamAdapter(String elementId) throws AdapterException {
 
     WorkerRestClient.stopStreamAdapter(ad.getSelectedEndpointUrl(), ad);
     ExtensionsLogProvider.INSTANCE.reset(elementId);
+
+    // remove the adapter from the metrics manager so that
+    // no metrics for this adapter are exposed anymore
+    AdapterMetricsManager.INSTANCE.getAdapterMetrics().remove(ad.getElementId(), ad.getName());
```
[the remark this answers is not in the document]
Comment (bossenti) [reply streampipes-connect-management/src/main/java/org/apache/streampipes/connect/management/management/AdapterMasterManagement.java:154]: Done

File: streampipes-commons/src/test/java/org/apache/streampipes/commons/prometheus/adapter/AdapterMetricsTest.java (hunk 1 above)
Comment (bossenti) [reply streampipes-commons/src/test/java/org/apache/streampipes/commons/prometheus/adapter/AdapterMetricsTest.java]: Done

**Comment by bossenti** [comment]
@tenthe please review this commit https://github.com/apache/streampipes/pull/2283/commits/61bf68f8e8b52c4d431eda99190f99d0e4958686 carefully

**Pull request merged**